lib/src/client.dart

```diff
@@ -95,34 +168,166 @@ class ClientConnection {
+    final stream = _transport.makeRequest(headers);
+    call._onConnectedStream(stream);
+  }
+
+  void _shutdownCall(ClientCall call) {
+    if (call._isCancelled) return;
+    call._onConnectError(
+        new GrpcError.unavailable('Connection shutting down.'));
+  }
+
   /// Shuts down this connection.
   ///
   /// No further calls may be made on this connection, but existing calls
   /// are allowed to finish.
   Future<Null> shutdown() {
-    // TODO(jakobr): Manage streams, close [_transport] when all are done.
-    return _transport.finish();
+    if (_state == ConnectionState.shutdown) return new Future.value();
+    _setShutdownState();
+    return _transport?.finish() ?? new Future.value();
   }
 
   /// Terminates this connection.
   ///
   /// All open calls are terminated immediately, and no further calls may be
   /// made on this connection.
   Future<Null> terminate() {
-    // TODO(jakobr): Manage streams, close them immediately.
-    return _transport.terminate();
+    _setShutdownState();
+    return _transport?.terminate() ?? new Future.value();
   }
 
-  /// Starts a new RPC on this connection.
-  ///
-  /// Creates a new transport stream on this connection, and sends initial call
-  /// metadata.
-  ClientTransportStream sendRequest(
-      bool useTls, String authority, String path, CallOptions options) {
-    final headers = createCallHeaders(useTls, authority, path, options);
-    final stream = _transport.makeRequest(headers);
-    // TODO(jakobr): Manage streams. Subscribe to stream state changes.
-    return stream;
+  void _setShutdownState() {
+    _setState(ConnectionState.shutdown);
+    _cancelTimer();
+    _pendingCalls.forEach(_shutdownCall);
+    _pendingCalls.clear();
+  }
+
+  void _setState(ConnectionState state) {
+    _state = state;
+    if (onStateChanged != null) {
+      onStateChanged(this);
+    }
+  }
+
+  void _handleIdleTimeout() {
+    if (_timer == null) return;
+    _timer = null;
+    _transport?.finish()?.catchError((_) => {}); // TODO(jakobr): Log error.
+    _transport = null;
+    _setState(ConnectionState.idle);
+  }
+
+  void _cancelTimer() {
+    _timer?.cancel();
+    _timer = null;
+  }
+
+  void _handleActiveStateChanged(bool isActive) {
+    if (isActive) {
+      _cancelTimer();
+    } else {
+      if (options.idleTimeout != null) {
+        _timer ??= new Timer(options.idleTimeout, _handleIdleTimeout);
+      }
+    }
+  }
+
+  bool _hasPendingCalls() {
+    // Get rid of pending calls that have timed out.
+    _pendingCalls.removeWhere((call) => call._isCancelled);
+    return _pendingCalls.isNotEmpty;
+  }
+
+  void _handleTransientFailure(error) {
+    _transport = null;
+    if (_state == ConnectionState.shutdown || _state == ConnectionState.idle) {
+      return;
+    }
+    // TODO(jakobr): Log error.
+    _cancelTimer();
+    if (!_hasPendingCalls()) {
+      // If there are no pending calls, just go directly to idle.
```

Comment:
[nit] This comment just repeats what the code already says.

Reply:
True, it does now. Removed.